chore(doc): improve doc (#115)
jenkins-infra/openvpn: README.md

* Open an INFRA ticket on [JIRA](https://issues.jenkins-ci.org) referencing your PR
* Grab a cup of coffee and wait patiently for an administrator to sign your certificate request
* Once an admin notify you that everything is right, you can [sync your fork](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) then pull it to retrieve your certificate from `./cert/pki/issued/<your-jenkins-username>.crt`
* We recommend you to move the generated files and the ca.cert to an hidden folder in your home (`~/.cert`):
  ```bash
  mkdir -p ~/.cert/jenkins-infra
  mv ./cert/pki/issued/<your-jenkins-username>.crt ~/.cert/jenkins-infra/<your-jenkins-username>.crt
  mv ./cert/pki/private/<your-jenkins-username>.key ~/.cert/jenkins-infra/<your-jenkins-username>.key
  cp ./cert/pki/ca.crt ~/.cert/jenkins-infra/ca.crt
  ```
* You can finally create the config file used by your VPN client.

Example here for [Tunnelblick](https://tunnelblick.net/), an OSX VPN client, opening this file from the Finder should launch it:  

>   _jenkins-infra.ovpn_
>   ```text

Reviewer: ````suggestion
  _jenkins-infra.ovpn_

  ```text
````

MD031/blanks-around-fences: Fenced code blocks should be surrounded by blank lines

  client
  remote vpn.jenkins.io 443
  ca "~/.cert/jenkins-infra/ca.crt"
  cert "~/.cert/jenkins-infra/<your-jenkins-username>.crt"
  key "~/.cert/jenkins-infra/<your-jenkins-username>.key"
  auth-user-pass
  dev tun
  proto tcp
  nobind
  auth-nocache
  script-security 2
  persist-key
  persist-tun
  user nobody
  group nobody
  ```

> #### With the [NetworkManager](https://wiki.archlinux.org/title/NetworkManager) client, **you must enable** the option `Use this connection only for resources on its network`

#### Windows only

If you want to use multiple VPN connections at the same time with OpenVPN, you have to install a new TAP adapter. This can be very easily by running *as Admin* the `C:\Program Files\TAP-Windows\bin\addtap.bat`. The [TAP-Windows](https://community.openvpn.net/openvpn/wiki/ManagingWindowsTAPDrivers) tool is installed in parallel with OpenVPN.

> 
> 

Reviewer: ```suggestion

```

MD012/no-multiple-blanks: Multiple consecutive blank lines [Expected: 1; Actual: 2]

```

### DNS Problems

If you are having issues connecting to resources behind the VPN, but the VPN appears to be working correctly, check your DNS settings.  Some providers seem to filter out requests to the zone.  To test, try `dig release.ci.jenkins.io`, you should get something like this:

<details><summary>dig output (click to expand)</summary>

```text
; <<>> DiG 9.10.6 <<>> release.ci.jenkins.io
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 13457
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1220
;; QUESTION SECTION:
;release.ci.jenkins.io.         IN      A

;; ANSWER SECTION:
release.ci.jenkins.io.  3600    IN      CNAME   private.aks.jenkins.io.
private.aks.jenkins.io. 3600    IN      A       10.0.2.5

;; Query time: 80 msec
;; SERVER: 192.168.1.254#53(192.168.1.254)
;; WHEN: Tue Oct 12 20:49:59 CEST 2021
;; MSG SIZE  rcvd: 92
```

</details>

To enable a different DNS provider only when connected to the VPN you can add the following to you OpenVPN config file:

```dhcp-option DNS 8.8.8.8```

> 
> 

Reviewer: ```suggestion

```
MD012/no-multiple-blanks: Multiple consecutive blank lines [Expected: 1; Actual: 2]

Feel free to follow the next action points:

* [Fork](https://help.github.com/articles/fork-a-repo/) this repository on your own Github account: [fork the repo](https://github.com/jenkins-infra/openvpn/fork)
* Clone your fork locally: `git clone https://github.com/<your-github-username>/openvpn && cd openvpn`
* Build EASYVPN binary by running one of the following commands depending on your operating system:
  * `make init_osx`
  * `make init_linux`
  * `make init_windows` then copy `./utils/easyvpn/easyvpn.exe` at the root of this repository
* Generate your private key and certificate request: `./easyvpn request <your-jenkins-username>`  
  Your private key will be generated in `./cert/pki/private`  
  > ### This key **must** remain **secret**!
* Create a new Pull Request on [jenkinsinfra/openvpn](https://github.com/jenkins-infra/openvpn), `staging` branch: [How to Create a pull request](https://help.github.com/articles/creating-a-pull-request/)
* Open an INFRA ticket on [JIRA](https://issues.jenkins-ci.org) referencing your PR
* Grab a cup of coffee and wait patiently for an administrator to sign your certificate request

> * Once an admin notify you that everything is right, you can [sync your fork](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) then pull it to retrieve your certificate from `./cert/pki/issued/<your-jenkins-username>.crt`

Reviewer: ```suggestion
* Once an admin notifies you that everything is setup, you can [sync your fork](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) then pull it to retrieve your certificate from `./cert/pki/issued/<your-jenkins-username>.crt`
```

* **username.key** cfr [HowTo Get client access](#howto-get-client-access) !! Your private key **must** remain **secret**,
* **username.crt** is located in [cert/pki/issued](https://github.com/jenkins-infra/openvpn/tree/master/cert/pki/issued), once an administrator signs  your request and publish it.

**With NetworkManager client, you must enable the option :**

`Use this connection only for resources on its network`

### DNS Problems

If you are having issues connecting to resources behind the VPN, but the VPN appears to be working correctly, check your DNS settings.  Some providers seem to filter out requests to the zone.  To test, try `dig release.ci.jenkins.io`, you should get something like this:

> <details><summary>dig output</summary>
> ```

Reviewer: ````suggestion

<details><summary>dig output (click to expand)</summary>

```text
````
- Since it's a markdown source file, you need to add empty lines between blocks to avoid breaking the final rendering
- Specifies a default language (`text`) for the fenced code block as per [MD040](https://github.com/DavidAnson/markdownlint/blob/main/doc/Rules.md#md040---fenced-code-blocks-should-have-a-language-specified)

; <<>> DiG 9.10.6 <<>> release.ci.jenkins.io
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 13457
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1220
;; QUESTION SECTION:
;release.ci.jenkins.io.         IN      A

;; ANSWER SECTION:
release.ci.jenkins.io.  3600    IN      CNAME   private.aks.jenkins.io.
private.aks.jenkins.io. 3600    IN      A       10.0.2.5

;; Query time: 80 msec
;; SERVER: 192.168.1.254#53(192.168.1.254)
;; WHEN: Tue Oct 12 20:49:59 CEST 2021
;; MSG SIZE  rcvd: 92
```

> </details>

Reviewer: ```suggestion

</details>
```
Since it's a markdown source file, you need to add empty lines between blocks to avoid breaking the final

Feel free to follow the next action points:

* [Fork](https://help.github.com/articles/fork-a-repo/) this repository on your own Github account: [fork the repo](https://github.com/jenkins-infra/openvpn/fork)
* Clone your fork locally: `git clone https://github.com/<your-github-username>/openvpn && cd openvpn`

> * Build EASYVPN binary by running one of the following commands depending on your OS:

Reviewer: ```suggestion
* Build EASYVPN binary by running one of the following commands depending on your operating system:
```

* Once an admin notify you that everything is right, you can [sync your fork](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) then pull it to retrieve your certificate from `./cert/pki/issued/<your-jenkins-username>.crt`
* We recommend you to move the `./cert` folder to an hidden folder in your home (`~/.cert`)
* You can finally create the config file used by your VPN client (example here for Tunnelblick, an OSX VPN client, opening this file from the Finder should launch it)  

_jenkins-infra.ovpn_
```text
client
remote vpn.jenkins.io 443
ca "~/.cert/pki/ca.crt"
cert "~/.cert/pki/issued/<your-jenkins-username>.crt"
key "~/.cert/pki/private/<your-jenkins-username>.key"
auth-user-pass
dev tun
proto tcp
nobind
auth-nocache
script-security 2
persist-key
persist-tun
user nobody
group nobody
```

#### HowTo show request information

* Enter in the VPN network directory: `cd cert`
* Run `make show-req name=<your-jenkins-username>`

#### HowTo show certificate information

* Install [sops](https://github.com/mozilla/sops)
* Enter in the VPN network directory: `cd ~/.cert`
* Run `make decrypt`

> * Run `make show-cert name=<your-jenkins-username>`

Reviewer: I realize, while reviewing, that this sub-section should be moved under the "Administrator" sub section, because the command `make decrypt` requires the user to be allowed to decrypt, which is only for admins...

WDYT about this? That would avoid polluting the mind of the reader who only want to get a VPN access)

Author: I've reorganized (and simplified some of) the first sections.

Reviewer: good work!

* Build EASYVPN binary by running one of the following commands depending on your
  * `make init_osx`
  * `make init_linux`
  * `make init_windows` then copy `./utils/easyvpn/easyvpn.exe` at the root of this repository
* Merge the Pull Request of the requester to the `staging` branch.
* Git checkout on the right branch `staging` to retrieve the [CRL](https://en.wikipedia.org/wiki/Certificate_revocation_list) from the requester
* Sign the certificate request: `./easyvpn sign <CN_to_sign>`
* Merge `staging` into `master`
* Update the Docker image tag in the [puppet](https://github.com/jenkins-infra/jenkins-infra/blob/staging/dist/profile/manifests/openvpn.pp) configuration.

### HowTo revoke client access?

* Build EASYVPN binary by running one of the following commands depending on your
  * `make init_osx`
  * `make init_linux`
  * `make init_windows` and copy `./utils/easyvpn/easyvpn.exe` at the root of this repository
* Revoke the certificate: `./easyvpn revoke <CN_to_sign>`
* Update the Docker image tag in the [puppet](https://github.com/jenkins-infra/jenkins-infra/blob/staging/dist/profile/manifests/openvpn.pp) configuration.

#### HowTo review certificate revocation list

> If the CLR expired, then the OpenVPN logs will contains errors like 'VERIFY ERROR: depth=0, error=CRL has expired:...'

Reviewer: ```suggestion
If the [CRL (Certificate Revocation list)](https://en.wikipedia.org/wiki/Certificate_revocation_list) expired, then the OpenVPN logs will contain errors like 'VERIFY ERROR: depth=0, error=CRL has expired:...'
```

Some examples can be found inside [docker-compose.yaml](docker/docker-compose.yaml)

### Testing

> To test this image, you need a "mock" ldap and SSL certificates, then go in the root folder and run `make start` to start the ldap and vpn service.
> > #### Certificates must be readable by UID 101!

Reviewer: ```suggestion
To test this image, you need a "mock" ldap and SSL certificates, then go in the root folder and run `make start` to start the ldap and vpn service.

> #### Certificates must be readable by UID 101!
```
MD022/blanks-around-headings/blanks-around-headers: Headings should be surrounded by blank lines [Expected: 1; Actual: 0; Above]
